extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/MessageBundleProcessor.java

```diff
@@ -955,193 +963,207 @@ private String generateImplementation(MessageBundleBuildItem bundle, ClassInfo d
 
         String targetPackage = DotNames.internalPackageNameWithTrailingSlash(bundleInterface.name());
         String generatedName = targetPackage + baseName + SUFFIX;
-
-        // MyMessages_Bundle implements MyMessages, Resolver
-        Builder builder = ClassCreator.builder().classOutput(classOutput).className(generatedName)
-                .interfaces(bundleInterface.name().toString(), Resolver.class.getName());
-        if (defaultBundleImpl != null) {
-            builder.superClass(defaultBundleImpl);
-        }
-        ClassCreator bundleCreator = builder.build();
-
-        // key -> method
-        Map<String, MessageMethod> keyMap = new LinkedHashMap<>();
-        List<MethodInfo> methods = new ArrayList<>(bundleInterfaceWrapper.methods());
-        // Sort methods
-        methods.sort(Comparator.comparing(MethodInfo::name).thenComparing(Comparator.comparing(MethodInfo::toString)));
-
-        for (MethodInfo method : methods) {
-            if (!method.returnType().name().equals(DotNames.STRING)) {
-                throw new MessageBundleException(
-                        String.format("A message bundle method must return java.lang.String: %s#%s",
-                                bundleInterface, method.name()));
+        String generatedClassName = generatedName.replace('/', '.');
+        String resolveMethodPrefix = baseName + SUFFIX;
+
+        Gizmo.create(classOutput).class_(generatedClassName, cc -> {
+            // MyMessages_Bundle implements MyMessages, Resolver
+            cc.implements_(ClassDesc.of(bundleInterface.name().toString()));
+            cc.implements_(Resolver.class);
+            if (defaultBundleImpl != null) {
+                cc.extends_(ClassDesc.of(defaultBundleImpl));
             }
-            LOG.debugf("Found message bundle method %s on %s", method, bundleInterface);
+            cc.defaultConstructor();
+
+            // key -> method
+            Map<String, MessageMethod> keyMap = new LinkedHashMap<>();
+            List<MethodInfo> methods = new ArrayList<>(bundleInterfaceWrapper.methods());
+            // Sort methods
+            methods.sort(Comparator.comparing(MethodInfo::name).thenComparing(Comparator.comparing(MethodInfo::toString)));
+
+            for (MethodInfo method : methods) {
+                cc.method(methodDescOf(method), mc -> {
+                    List<ParamVar> params = new ArrayList<>(method.parametersCount());
+                    for (MethodParameterInfo param : method.parameters()) {
+                        params.add(mc.parameter(param.name() != null ? param.name() : "arg" + param.position(),
+                                classDescOf(param.type())));
```

Comment:
Since you're creating the method from the full descriptor, you no longer have to provide types, just names and positions.

(FTR, I'm planning to add `MethodParameterInfo.nameOrDefault()`, because I too got bored by these constant `null` checks. Alas, it's not there yet.)